### ISSUE TITLE
Adding PyFoam template syntax in OpenFOAM file crashes Neovim  

### ISSUE THREAD

**Issue by evanhorn** [nvim-treesitter/nvim-treesitter#2264]
### Before reporting

- [X] I have updated my neovim version to latest _master_
- [X] I have updated my plugin to the latest version
- [X] I have run `:TSUpdate`
- [X] I have read the [troubleshooting section](https://github.com/nvim-treesitter/nvim-treesitter#troubleshooting)

### Describe the bug

I was editing an OpenFoam (actually a template file for use with pyFoam), and when I typed in ```<!--(``` (to add the script), I got a segmentation fault and Neovim crashed. A couple of notes, the crash:

* does not occur when highlighting is disabled.
* does not occur with a blank file with the filetype set to "foam".
* seems to be dependent on where the characters are typed into the file (see below). 

I have generated a minimal config file and tested it on the fvSchemes file for one of the tutorials ($FOAM_TUT/incompressible/icoFoam/cavity/cavity/system/fvSchemes).

Most of the time I only get the notice ```Segmentation fault (core dumped)```, but occasionally I get the message ```*** stack smashing detected ***: terminated```.

Using $FOAM_TUTORIALS/incompressible/icoFoam/cavity/cavity/system/fvSchemes as an example, typing ```<!--(``` above the laplacianSchemes dictionary crashes Neovim, but typing it below that dictionary does not cause a crash. Furthermore, within the laplacianSchemes dictionary, the crash does not occur if typed below the "default" scheme, but does crash if typed above it.

I have found that the same problem exists with other files as well, e.g., controDict, but I have not tested it on all the files. I hope that this will be sufficient to determine the source of the problem.

### To Reproduce

1. cp -r $FOAM_TUTORIAL/incompressible/icoFoam/cavity/cavity $FOAM_RUN
2. vi $FOAM_RUN/cavity/system/fvSchemes
3. Type ```<!--(``` directly above 'gradSchemes'

### Expected behavior

Program does not crash.

### Output of `:checkhealth nvim-treesitter`

```markdown
nvim-treesitter: require("nvim-treesitter.health").check()
========================================================================
## Installation
  - WARNING: `tree-sitter` executable not found (parser generator, only needed for :TSInstallFromGrammar, not required for :TSInstall)
  - OK: `node` found v17.0.1 (only needed for :TSInstallFromGrammar)
  - OK: `git` executable found.
  - OK: `cc` executable found. Selected from { vim.NIL, "cc", "gcc", "clang", "cl", "zig" }
    Version: cc (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0
  - OK: Neovim was compiled with tree-sitter runtime ABI version 13 (required >=13). Parsers must be compatible with runtime ABI.

## Parser/Features H L F I J
  - ocaml_interface✓ ✓ ✓ . ✓ 
  - beancount      ✓ . ✓ . . 
  - python         ✓ ✓ ✓ ✓ ✓ 
  - sparql         ✓ ✓ ✓ ✓ ✓ 
  - comment        ✓ . . . . 
  - lua            ✓ ✓ ✓ ✓ ✓ 
  - ocaml          ✓ ✓ ✓ . ✓ 
  - ninja          ✓ . ✓ ✓ . 
  - ql             ✓ ✓ . ✓ ✓ 
  - glsl           ✓ ✓ ✓ ✓ ✓ 
  - json           ✓ ✓ ✓ ✓ . 
  - pug            ✓ . . . ✓ 
  - ledger         ✓ . ✓ ✓ ✓ 
  - php            ✓ ✓ ✓ ✓ ✓ 
  - clojure        ✓ ✓ ✓ . ✓ 
  - supercollider  ✓ ✓ ✓ ✓ ✓ 
  - heex           ✓ . ✓ ✓ ✓ 
  - typescript     ✓ ✓ ✓ ✓ ✓ 
  - fennel         ✓ ✓ . . ✓ 
  - query          ✓ ✓ ✓ ✓ ✓ 
  - cpp            ✓ ✓ ✓ ✓ ✓ 
  - vue            ✓ . ✓ ✓ ✓ 
  - latex          ✓ . ✓ . ✓ 
  - make           ✓ . . . ✓ 
  - rst            ✓ ✓ . . ✓ 
  - css            ✓ . ✓ ✓ ✓ 
  - glimmer        ✓ . . . . 
  - erlang         . . . . . 
  - gowork         ✓ . . . ✓ 
  - regex          ✓ . . . . 
  - svelte         ✓ . ✓ ✓ ✓ 
  - nix            ✓ ✓ ✓ . ✓ 
  - elm            ✓ . . . ✓ 
  - c              ✓ ✓ ✓ ✓ ✓ 
  - fortran        ✓ . ✓ ✓ . 
  - java           ✓ ✓ . ✓ ✓ 
  - prisma         ✓ . . . . 
  - http           ✓ . . . ✓ 
  - hjson          ✓ ✓ ✓ ✓ ✓ 
  - gomod          ✓ . . . ✓ 
  - dart           ✓ ✓ . ✓ ✓ 
  - verilog        ✓ ✓ ✓ . ✓ 
  - llvm           ✓ . . . . 
  - pioasm         ✓ . . . ✓ 
  - bash           ✓ ✓ ✓ . ✓ 
  - json5          ✓ . . . ✓ 
  - vim            ✓ ✓ . . ✓ 
  - yaml           ✓ ✓ ✓ ✓ ✓ 
  - cmake          ✓ . ✓ . . 
  - zig            ✓ . ✓ ✓ ✓ 
  - bibtex         ✓ . ✓ ✓ . 
  - r              ✓ ✓ . . . 
  - julia          ✓ ✓ ✓ ✓ ✓ 
  - jsdoc          ✓ . . . . 
  - turtle         ✓ ✓ ✓ ✓ ✓ 
  - pascal         ✓ ✓ ✓ ✓ ✓ 
  - toml           ✓ ✓ ✓ ✓ ✓ 
  - tsx            ✓ ✓ ✓ ✓ ✓ 
  - fusion         ✓ ✓ ✓ ✓ . 
  - javascript     ✓ ✓ ✓ ✓ ✓ 
  - yang           ✓ . ✓ . . 
  - markdown       ✓ . ✓ . ✓ 
  - surface        ✓ . ✓ ✓ ✓ 
  - haskell        ✓ . . . ✓ 
  - dot            ✓ . . . ✓ 
  - go             ✓ ✓ ✓ ✓ ✓ 
  - commonlisp     ✓ ✓ ✓ . . 
  - cuda           ✓ ✓ ✓ ✓ ✓ 
  - c_sharp        ✓ ✓ ✓ . ✓ 
  - elixir         ✓ ✓ ✓ ✓ ✓ 
  - kotlin         ✓ ✓ ✓ . ✓ 
  - ruby           ✓ ✓ ✓ ✓ ✓ 
  - fish           ✓ ✓ ✓ ✓ ✓ 
  - scala          ✓ . ✓ . ✓ 
  - dockerfile     ✓ . . . ✓ 
  - rust           ✓ ✓ ✓ ✓ ✓ 
  - jsonc          ✓ ✓ ✓ ✓ ✓ 
  - perl           ✓ . . . . 
  - tlaplus        ✓ . ✓ . ✓ 
  - html           ✓ ✓ ✓ ✓ ✓ 
  - scss           ✓ . . ✓ . 
  - graphql        ✓ . . ✓ ✓ 
  - foam           ✓ ✓ ✓ ✓ ✓ 
  - rasi           ✓ ✓ ✓ ✓ . 
  - hcl            ✓ . ✓ ✓ ✓ 

  Legend: H[ighlight], L[ocals], F[olds], I[ndents], In[j]ections
         +) multiple parsers found, only one will be used
         x) errors found in the query, try to run :TSUpdate {lang}
```

### Output of `nvim --version`

```text
NVIM v0.7.0-dev
Build type: RelWithDebInfo
LuaJIT 2.1.0-beta3
Compilation: /usr/bin/cc -g -O2 -fdebug-prefix-map=/build/neovim-oR9fFZ/neovim-0.7.0~ubuntu1+git202201130047-39a35dd00-dd05b3569=. -fstack-protector-strong -Wformat -Werror=format-security -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1 -DNVIM_TS_HAS_SET_MATCH_LIMIT -DNVIM_TS_HAS_SET_ALLOCATOR -O2 -g -Og -g -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wshadow -Wconversion -Wmissing-prototypes -Wimplicit-fallthrough -Wvla -fstack-protector-strong -fno-common -fdiagnostics-color=auto -DINCLUDE_GENERATED_DECLARATIONS -D_GNU_SOURCE -DNVIM_MSGPACK_HAS_FLOAT32 -DNVIM_UNIBI_HAS_VAR_FROM -DMIN_LOG_LEVEL=3 -I/build/neovim-oR9fFZ/neovim-0.7.0~ubuntu1+git202201130047-39a35dd00-dd05b3569/build/config -I/build/neovim-oR9fFZ/neovim-0.7.0~ubuntu1+git202201130047-39a35dd00-dd05b3569/src -I/build/neovim-oR9fFZ/neovim-0.7.0~ubuntu1+git202201130047-39a35dd00-dd05b3569/.deps/usr/include -I/usr/include -I/build/neovim-oR9fFZ/neovim-0.7.0~ubuntu1+git202201130047-39a35dd00-dd05b3569/build/src/nvim/auto -I/build/neovim-oR9fFZ/neovim-0.7.0~ubuntu1+git202201130047-39a35dd00-dd05b3569/build/include
Compiled by buildd@lcy02-amd64-004

Features: +acl +iconv +tui
See ":help feature-compile"

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "/usr/share/nvim"

Run :checkhealth for more info
```

### Additional context

Minimal config file:

```
call plug#begin(stdpath('data').'/plugged')

Plug 'nvim-treesitter/nvim-treesitter', {'do': ':TSUpdate'}

call plug#end()

lua << EOF
require("nvim-treesitter.configs").setup {
  ensure_installed = "all", -- one of "all", "maintained" (parsers with maintainers), or a list of languages
  highlight = {
    enable = true,
  },
}
EOF
```

**Comment by theHamsta** [comment]
Can you report this at the upstream parser, too? @FoamScience can you please also experiment a bit with fuzzing https://github.com/ikatyang/tree-sitter-markdown/issues/14 explanations are still up-to-date except that you can now also use Python 3

**Comment by FoamScience** [comment]
I'm on it!

**Comment by FoamScience** [comment]
@evanhorn I'm glad you're already using the parser! PR #2274 fixes this issue (Can you please confirm by checking out the PR, and `:TSUpdate foam` ?).
Also, I would be interested in any feedback from you if you can [testOFFiles.sh](https://github.com/FoamScience/tree-sitter-foam/blob/master/testOFFiles.sh) on your run directory.
Plus, please fill subsequent issues/questions at [tree-sitter-foam](https://github.com/FoamScience/tree-sitter-foam) if you think they are not related to `nvim-treesitter` but to the parser itself 😄 .

@theHamsta This was a bug in the scanner, I didn't need to run fuzzers to figure it out, but their output is interesting.
With leaks, I got 3 16 bytes leaks in `ts_parser_set_language`; I guess this has nothing to do with the parser:
```

Direct leak of 16 byte(s) in 1 object(s) allocated from:
    #0 0x524df2 in calloc (/tmp/tree-sitter/out/foam_fuzzer+0x524df2)
    #1 0x586cad in ts_parser_set_language (/tmp/tree-sitter/out/foam_fuzzer+0x586cad)
    #2 0x5590df in LLVMFuzzerTestOneInput (/tmp/tree-sitter/out/foam_fuzzer+0x5590df)
    #3 0x45e861 in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) (/tmp/tree-sitter/out/foam_fuzzer+0x45e861)
    #4 0x45e0a5 in fuzzer::Fuzzer::RunOne(unsigned char const*, unsigned long, bool, fuzzer::InputInfo*, bool*) (/tmp/tree-sitter/out/foam_fuzzer+0x45e0a5)
    #5 0x4608da in fuzzer::Fuzzer::ReadAndExecuteSeedCorpora(std::Fuzzer::vector<fuzzer::SizedFile, fuzzer::fuzzer_allocator<fuzzer::SizedFile> >&) (/tmp/tree-sitter/out/foam_fuzzer+0x4608da)
    #6 0x460cd9 in fuzzer::Fuzzer::Loop(std::Fuzzer::vector<fuzzer::SizedFile, fuzzer::fuzzer_allocator<fuzzer::SizedFile> >&) (/tmp/tree-sitter/out/foam_fuzzer+0x460cd9)
    #7 0x44ee48 in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) (/tmp/tree-sitter/out/foam_fuzzer+0x44ee48)
    #8 0x478292 in main (/tmp/tree-sitter/out/foam_fuzzer+0x478292)
    #9 0x7f7c922d90b2 in __libc_start_main /build/glibc-eX1tMB/glibc-2.31/csu/../csu/libc-start.c:308:16
```

`./out/foam_fuzzer -detect_leaks=0` completes without problems.
```
#89011	NEW    cov: 6779 ft: 36614 corp: 1796/12614b lim: 17 exec/s: 209 rss: 450Mb L: 17/17 MS: 1 ChangeBit-
#89038	NEW    cov: 6779 ft: 36615 corp: 1797/12631b lim: 17 exec/s: 209 rss: 450Mb L: 17/17 MS: 2 CrossOver-CrossOver-
#89110	REDUCE cov: 6779 ft: 36615 corp: 1797/12629b lim: 17 exec/s: 209 rss: 450Mb L: 6/17 MS: 2 PersAutoDict-EraseBytes- DE: "\x01\x00\x00\x1c"-
#89218	NEW    cov: 6779 ft: 36621 corp: 1798/12646b lim: 17 exec/s: 209 rss: 450Mb L: 17/17 MS: 3 InsertByte-ShuffleBytes-CrossOver-
#89223	NEW    cov: 6779 ft: 36628 corp: 1799/12663b lim: 17 exec/s: 209 rss: 450Mb L: 17/17 MS: 5 InsertRepeatedBytes-PersAutoDict-ChangeBinInt-ChangeBinInt-CrossOver- DE: "\x01\x00\x00\x02"-
#89274	NEW    cov: 6779 ft: 36649 corp: 1800/12680b lim: 17 exec/s: 209 rss: 450Mb L: 17/17 MS: 1 CopyPart-
#89311	NEW    cov: 6779 ft: 36654 corp: 1801/12697b lim: 17 exec/s: 209 rss: 450Mb L: 17/17 MS: 2 ChangeByte-CrossOver-
#89342	REDUCE cov: 6779 ft: 36654 corp: 1801/12696b lim: 17 exec/s: 209 rss: 450Mb L: 10/17 MS: 1 EraseBytes-
#89458	NEW    cov: 6779 ft: 36655 corp: 1802/12709b lim: 17 exec/s: 209 rss: 451Mb L: 13/17 MS: 1 CrossOver-
#89459	NEW    cov: 6779 ft: 36656 corp: 1803/12721b lim: 17 exec/s: 209 rss: 451Mb L: 12/17 MS: 1 EraseBytes-
#89491	NEW    cov: 6779 ft: 36658 corp: 1804/12730b lim: 17 exec/s: 209 rss: 451Mb L: 9/17 MS: 2 InsertByte-ChangeBinInt-
#89492	NEW    cov: 6779 ft: 36660 corp: 1805/12747b lim: 17 exec/s: 209 rss: 451Mb L: 17/17 MS: 1 CopyPart-
#89503	NEW    cov: 6779 ft: 36664 corp: 1806/12764b lim: 17 exec/s: 209 rss: 451Mb L: 17/17 MS: 1 CrossOver-
'
```

**Comment by theHamsta** [comment]
I assume this has be solved. @evanhorn can you please re-open if this issue persists?

**Comment by evanhorn** [comment]
Sorry for the delay — I was AFK for the past few days — but the fix appears to work great. Thanks for the quick response!

@FoamScience, thanks for your work on this. I was using a VIM syntax file that I modified for OpenFOAM8, and found yours when I decided to see if anyone had developed a Treesitter version. I am currently trying to teach myself OpenFOAM through the ["3 weeks" series](https://wiki.openfoam.com/index.php?title=%223_weeks%22_series).
